openslides_backend/action/actions/user/save_saml_account.py

```diff
@@ -0,0 +1,131 @@
+                "SingleSignOn field attributes are not configured in OpenSlides"
+            )
+        new_instance = {
+            attr_config[key]: value
+            for key, value in instance.items()
+            if key in attr_config
+            and attr_config[key] in self.schema["properties"].keys()
+        }
+        instance.clear()
+        instance.update(new_instance)
+        if "saml_id" not in instance:
+            raise ActionException(
+                "There is no field of user's data mapped to the required 'saml_id'"
+            )
+        super().validate_instance(instance)
+
+    def prefetch(self, action_data: ActionData) -> None:
+        if len(cast(Sized, action_data)) != 1:
+            raise ActionException(
+                "The save_saml_account action accepts only one user instance!"
+            )
+
+    def prepare_action_data(self, action_data: ActionData) -> ActionData:
+        for instance in action_data:
+            users = self.datastore.filter(
+                "user", FilterOperator("saml_id", "=", instance["saml_id"]), ["id"]
+            )
+            if len(users) == 1:
+                instance["id"] = next(iter(users.values()))["id"]
+            elif len(users) == 0:
+                new_id = self.datastore.reserve_ids(self.model.collection, 1)[0]
+                instance["id"] = new_id
+                instance["can_change_own_password"] = False
+                instance["organization_id"] = ONE_ORGANIZATION_ID
+                instance["username"] = self.generate_usernames([instance["saml_id"]])[0]
+                instance["meta_new"] = True
+            else:
+                ActionException(
+                    f"More than one existing user found in database with saml_id {instance['aml_id']}"
+                )
+        return action_data
+
+    def check_permissions(self, instance: Dict[str, Any]) -> None:
+        pass
+
+    def base_update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
+        if "meta_new" in instance:
+            instance = self.set_defaults(instance)
+        instance = self.validate_fields(instance)
+
+        instance = self.update_instance(instance)
+        self.apply_instance(instance)
+        self.validate_relation_fields(instance)
+
+        return instance
+
+    def create_events(self, instance: Dict[str, Any]) -> Iterable[Event]:
+        """
+        Handles create and update
+        """
+        fqid = fqid_from_collection_and_id(self.model.collection, instance["id"])
+        if "meta_new" in instance:
+            del instance["meta_new"]
+            yield self.build_event(EventType.Create, fqid, instance)
+        else:
+            fields = {
+                k: v
+                for k, v in instance.items()
+                if k != "id" and not k.startswith("meta_")
+            }
+            if not fields:
+                return []
+            yield self.build_event(EventType.Update, fqid, fields)
```

Comment:
You can inherit from both `CreateAction` and `UpdateAction` and call the respective `super(SuperClass).create_events()` methods here to reduce code duplication

Reply:
or, alternatively, move the code to a mixin and use it here and in `UpdateAction`